Bloodhound

huangwaylon opened the pull request:
This branch adds the bloodhound component, as described by the GRT paper, into Randoop.

Specifically, it weights each method under test based on the method's branch coverage and the number of times the method is chosen for a new sequence.

mernst commented:
@huangwaylon Lots of tests are failing.  Do you understand why?

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -38,6 +41,39 @@
 /** Randoop's forward, component-based generator. */
 public class ForwardGenerator extends AbstractGenerator {
 
+  /**
```
This is a lot of fields, with unclear documentation (such as just "Hyperparameter for decreasing weights.") and undocumented relationship to the rest of the class.  Maybe the structure could be clarified by making a struct for all of these?  That could also make it easier to turn on and off the entire feature.

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -151,6 +195,99 @@ public ExecutableSequence step() {
   }
 
+  /**
+   * Recompute weights for all methods under test at regular intervals.
```
I don't see anywhere that the documentation references GRT.  Without documentation of the intended effect, it is difficult for readers to understand this or to validate that it works as intended.

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -151,6 +195,99 @@ public ExecutableSequence step() {
+    // After the specified interval, recompute the current coverage information.
+    if (stepNum % interval == 0) {
+
+      // Print method weights.
```
What is the purpose of this?  Won't it clutter the output?

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -151,6 +195,99 @@ public ExecutableSequence step() {
+   * @param interval interval at which to recompute weights
+   */
+  private void processWeightsForMethods(int interval) {
+    // Increment our step counter.
```
It's not helpful to give comments that restate the exact operation.  Remove them for brevity.  There are a number of examples like this; another is "clear the method selections map".

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -275,7 +412,22 @@ private ExecutableSequence createNewUniqueSequence() {
+    }
+
+    // Update the number of times this method was selected for a new sequence.
+    incrementInMap(methodSelections, operation);
```
If this is only used in bloodhound, should it also be in the "if" statement?  Or, abstract this all into a method that updates values and returns a random member.

mernst commented on src/main/java/randoop/main/GenInputsAbstract.java:
```diff
@@ -516,6 +516,10 @@ public Limits(int time_limit, int attempted_limit, int generated_limit, int outp
     ALL
   }
 
+  /** Enable blood hound to prioritize methods with lower branch coverage. */
```
Throughout, "bloodhound" is a single word.

mernst commented on src/main/java/randoop/main/GenInputsAbstract.java:
```diff
@@ -516,6 +516,10 @@ public Limits(int time_limit, int attempted_limit, int generated_limit, int outp
   }
 
+  /** Enable blood hound to prioritize methods with lower branch coverage. */
+  @Option("Enable blood hound to prioritize methods with lower branch coverage.")
```
I would make this an unadvertised option.

mernst commented on src/main/java/randoop/main/GenTests.java:
```diff
@@ -530,6 +530,9 @@ public boolean handle(String[] args) {
     // so dump after all done.
     explorer.getOperationHistory().outputTable();
 
+    // Shutdown coverage tracking object.
```
As a verb, "shut down" is two words.

mernst commented on src/main/java/randoop/reflection/OperationModel.java:
```diff
@@ -470,6 +471,7 @@ private void addClassTypes(
     Set<Class<?>> visitedClasses = new LinkedHashSet<>(); // consider each class just once
     for (String classname : classnames) {
       Class<?> c = getClass(classname, errorHandler);
+      c = CoverageTracker.instance.getInstrumentedClass(c != null ? c.getName() : "");
```
`getInstrumentedClass` returns null if its argument is `""`, but I had to look it up to learn that.  This would be clearer as
```
      c = (c != null ? CoverageTracker.instance.getInstrumentedClass(c.getName()) : null;
```
or
```
      if (c != null) {
        c = CoverageTracker.instance.getInstrumentedClass(c.getName());
      }
```

huangwaylon commented:
The errors are occurring in Randoop's coverage instrumentation tests.
The specific error that is common among all cases appears to be:
java.lang.IllegalAccessError: org/jacoco/core/instr/Instrumenter$1

I wonder if I need to / can exclude the project jacoco.core from this package through the build.gradle file.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  /**
+   * Recompute weights for all methods under test at regular intervals. Each method under test is
```
This documentation can be improved.  This method doesn't do anything at regular intervals.  Is it called at regular intervals?  Does this method always do recomputation, or just sometimes?  Please improve the documentation.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,187 @@
+   * ratio between the number of times this method has been invoked and the maximum number of times
+   * any method under test has been invoked. The weighting scheme is based on the scheme described
+   * by the authors of the Guided Random Testing (GRT) paper. We have chosen our own reasonable
+   * values for all parameters that were left unspecified by the authors of GRT.
```
Should we do experiments with other values?

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -92,6 +95,9 @@ public ForwardGenerator(
     initializeRuntimePrimitivesSeen();
+
+    // Copy all operations into bloodhound's list of operations.
+    bloodhound.setOperations(operations);
```
For better encapsulation, I suggest doing no initialization on line 73, and `bloodhound = new Bloodhound(operations);` here.

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -275,7 +281,17 @@ private ExecutableSequence createNewUniqueSequence() {
+    TypedOperation operation;
+
+    // If bloodhound is enabled, choose the next operation while considering the methods' weights.
+    if (GenInputsAbstract.enable_bloodhound) {
```
I suggest generalizing this.
Create an interface with method `getNextOperation`.
Bloodhound will implement the interface, and so will a trivial class (named, say, `UniformRandomMethodSelection`) that just calls `Randomness.randomMember(this.operations);`.

In this class, set a variable of the interface type, to be either an instance of Bloodhound or of UniformRandom (depending on `GenInputsAbstract.enable_bloodhound`).  Then, here there doesn't have to be any special-casing for bloodhound.

mernst commented on src/main/java/randoop/reflection/OperationModel.java:
```diff
@@ -466,10 +467,14 @@ private void addClassTypes(
       mgr.add(new ClassLiteralExtractor(this.classLiteralMap));
     }
 
+    // Need to load in instrumented versions of all classes under test first.
```
Should the changes in this class depend on `GenInputsAbstract.enable_bloodhound`, or should they be unconditional as they are now?

mernst commented on src/main/java/randoop/operation/TypedOperation.java:
```diff
@@ -613,4 +615,9 @@ public ExpectedOutcomeTable checkPrestate(Object[] values) {
     this.execSpec = execSpec;
   }
+
+  @Override
```
This is code I would encapsulate in the `UniformRandomMethodSelection` class.

mernst commented on src/main/java/randoop/main/CoverageTracker.java:
```diff
@@ -0,0 +1,192 @@
+    try {
+      runtime.startup(data);
+    } catch (Exception e) {
+      e.printStackTrace();
```
Should this terminate execution too?  Otherwise it's likely that a user won't notice the logging.

mernst commented on src/main/java/randoop/main/CoverageTracker.java:
```diff
@@ -0,0 +1,192 @@
+    return instrumentedClass;
+  }
+
+  /** Collect coverage information for all methods under test. Updates the coverageDetailsMap. */
```
Clarify whether this generates data (maybe by executing tests), or summarizes already-collected data, or prepares for later collection, or something else.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,206 @@
+    // Update the number of times this method was selected for a new sequence.
+    incrementInMap(methodSelections, operation);
+
+    // Update the number of times this method was successfully invoked.
```
This line speaks of "successfully invoked", but just above it's "selected".  These must be the same concept, because one is incremented whenever the other is.  Why are two different terms used for it?

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,200 @@
+   * Coverage details related to a single method under test. Tracks total number of branches and
+   * number of uncovered branches.
+   */
+  public static class CoverageDetails {
```
I suggest calling this `BranchCoverage` instead of `CoverageDetails`, because it is more descriptive.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,200 @@
+    if (!classesUnderTest.contains(className)) {
+      boolean isNestedClass = false;
+
+      // Determine if this class is a nested class, in which case, the name of some
```
I suggest abstracting `isNestedClassOf` into a separate method for clarity.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,200 @@
+      // Determine if this class is a nested class, in which case, the name of some
+      // existing class under test would be the prefix of this class's name.
+      for (String classUnderTest : classesUnderTest) {
+        if (className.startsWith(classUnderTest)) {
```
This is not enough.  The next character needs to be `$` as well.
So you could test for `startsWith(classUnderTest + "$")`.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,200 @@
+    InputStream original = getClass().getResourceAsStream(resource);
+
+    if (original == null) {
+      System.err.println("No resource with name: " + resource + " found!");
```
Can you be more explicit?
Is this a guaranteed precursor to a failure later?
Whose fault is it, and how could it be corrected?
Should Randoop halt?  Randoop is so noisy that the user is likely to overlook this output.

huangwaylon replied:
The only place where `instrumentClass` in `CoverageTracker` is invoked is from `MemoryClassLoader`'s `loadClass` method where it checks the returned result.
```
final byte[] bytes = coverageTracker.instrumentClass(name);
if (bytes != null) {
  loadedClass = defineClass(name, bytes, 0, bytes.length);
} else {
  loadedClass = super.loadClass(name, resolve);
}
```
In this case, the parent ClassLoader would attempt to load the class.
However, reaching this point would mean MemoryClassLoader wasn't able to find a class that was explicitly under test and thus the parent ClassLoader probably won't find it either. 
So throwing an error in CoverageTracker seems to be the right thing to do and to halt Randoop.

mernst replied:
OK.  So it sounds like you will change the `println` to `throw new Error`?

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,200 @@
+   *     null if class with target name cannot be found.
+   */
+  public Class<?> getInstrumentedClass(String className) {
+    if (className == null || className.isEmpty()) {
```
Should clients pass `""`?  It's not documented as a possibility.  I would drop the disjunct that tests for it.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,200 @@
+        analyzer.analyzeClass(original, className);
+        original.close();
+      } catch (IOException e) {
+        e.printStackTrace();
```
Why not `throw new Error(e)`?

mernst commented on src/main/java/randoop/reflection/OperationModel.java:
```diff
@@ -469,7 +470,13 @@ private void addClassTypes(
     for (String classname : classnames) {
-      Class<?> c = getClass(classname, errorHandler);
+      Class<?> c;
+      if (GenInputsAbstract.enable_bloodhound) {
```
Why isn't this code in `getClass`?

mernst commented on src/main/java/randoop/reflection/OperationModel.java:
```diff
@@ -502,7 +510,12 @@ private void addClassTypes(
       if (!classnames.contains(classname)) {
-        Class<?> c = getClass(classname, errorHandler);
+        Class<?> c;
+        if (GenInputsAbstract.enable_bloodhound) {
```
Why isn't this code in `getClass`?

huangwaylon commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,206 @@
+  private final double alpha = 0.7;
+
+  /** Hyper-parameter for decreasing weights of methods between updates to coverage information. */
+  private final double p = 0.5;
```
It is declared as "p" in the paper. I've added to the description to document this.

huangwaylon commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,206 @@
+  /** Hyper-parameter for decreasing weights of methods between updates to coverage information. */
+  private final double p = 0.5;
+
+  /** Hyper-parameter for determining when to recompute branch coverage. */
```
![image](https://user-images.githubusercontent.com/15660492/38319215-204657c6-37e6-11e8-9b18-7db86a433cef.png)
There doesn't appear to be a way to interpret this variable with a unit. "p" is used to compute some upper-bound by which we scale the weight of an element.

huangwaylon commented:
I was able to reproduce the fibHeap assertion error locally.
I added a line in AbstractGenerator's createAndClassifySequences method and it turned out that at the end, in a single step, it produced structures that covered two unique method branch fingerprints (As you suspected).
Added the change to use <= for number of branch fingerprints.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,199 @@
+   *     null if class with target name cannot be found.
+   */
+  public Class<?> getInstrumentedClass(String className) {
+    if (className == null) {
```
This behavior is not documented in the method specification.
Looking at the calls, they don't seem to ever pass null.  I think it would be better to remove this so that the code does not silently continue when there is a problem.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,199 @@
+    try {
+      instrumentedClass = memoryClassLoader.loadClass(className);
+    } catch (ClassNotFoundException e) {
+      e.printStackTrace();
```
Why does this print a stack trace but do no more?  If the problem is serious, I think Randoop should halt.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,199 @@
+   * contain the updated coverage information of each method branch.
+   */
+  public void collect() {
+    // Collect coverage information.
```
I'm not sure what "collect coverage information" means. The phrase appears here, and in the method Javadoc, and on line 165, but it must mean different things in all three places.  Please make each one more precise.

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,199 @@
+    CoverageBuilder coverageBuilder = new CoverageBuilder();
+    Analyzer analyzer = new Analyzer(executionData, coverageBuilder);
+
+    // Analyze the coverage of each of the tracked classes.
```
This comment is not very descriptive.  It's the same as the method name, so it doesn't add much beyond that.  It would be better to say what the method actually does.

mernst commented on src/main/java/randoop/generation/MemoryClassLoader.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.HashMap;
+import java.util.Map;
+
+/** A class loader that loads classes from in-memory data. */
```
This documentation does not say anything about instrumentation, but that is what the class really does.  The class should perhaps be renamed as well.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,210 @@
+  public TypedOperation selectOperation() {
+    // Collect branch coverage and recompute weights for methods under test.
+    updateBranchCoverageMaybe();
+    updateWeightsForOperations();
```
Why is updateWeightsForOperations called on every step?  It's called in selectOperation, right after updateBranchCoverageMaybe which only has an effect sometimes.
My first concern was the efficiency cost of re-computing weights every operation.
Also, the GRT paper says, "At each update of the coverage, the weights are recalculated, and k is reset to 0."  This implies that weights aren't recomputed at every step.

Should you replace

```
    updateBranchCoverageMaybe();
    updateWeightsForOperations();
```

by

```
    stepNum++;
    if (stepNum % branchCoverageInterval == 0) {
      updateBranchCoverage();
      updateWeightsForOperations();
    }
```

?

huangwaylon replied:
Hi Mike, 
I'm wondering if I'm misinterpreting this segment, but I do believe that weights are regularly updated at each step.
![image](https://user-images.githubusercontent.com/15660492/38590653-6fa7efd6-3ce7-11e8-95dc-970c59da2143.png)
In this segment, the authors define `w` as a function of the method and `k`. `k` is the "the number of selections of method m since the last update of the coverage information". Thus, it appears to be the case that a method's weight will change within the middle of an interval (where coverage is computed at the ends of the interval). 

In regards to efficiency, at the moment, I've implemented it such that the weights of ALL methods under test are recomputed at every step. I wonder if it's sufficient to only update the weight of the method that gets chosen to be used to create a "new and unique sequence".

huangwaylon replied:
And then, at each point where coverage information is updated/re-summarized, we would then recompute the weights of all the methods under test.

mernst replied:
Thanks for the clarification.  I had interpreted k as the number of selections of method m between updates to coverage information (that is, between the previous two coverage updates).  Your interpretation may be right -- it fits the text even better.

Updating only the weight of the method that gets chosen is sufficient and probably cheaper, since there are many methods under test.  (But, I haven't run timing tests.)  Could you try that?

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,210 @@
+   * during a run of Randoop. The GRT paper does not state its definition of the "number of
+   * invocations" of a method under test.
+   */
+  private final Map<TypedOperation, Integer> methodSuccessfulCalls = new HashMap<>();
```
methodSelections and methodSuccessfulCalls represent the same thing.  Why is does one of them have the name "...Selections" and the other has the name "...Calls"?  Please make this consistent by changing one or both names.

huangwaylon replied:
I had originally intended for the descriptions of the two variables in the GRT paper to be reflected in the variables' names.
![image](https://user-images.githubusercontent.com/15660492/38591412-7530a732-3ceb-11e8-93c3-cd9063767bc3.png)
However, I do agree that it would make more sense to rename the variables.
What do you think about:
`methodSelections` -> `currMethodSelectionCounts`
`methodSuccessfulCalls` -> `methodSelectionTotals`

mernst replied:
Thanks.  That text is very helpful.

I would guess that "selections" and "successful invocations" mean different things.  Here is my interpretation:
"Selections" means an attempt by Randoop to generate a test that culminates with the given method.
"Successful invocations" means that the call was classified as a normal execution (as opposed to being classified as fault-revealing or as invalid.

Do you think this is what the authors meant, or they were using the two distinct terms as synonyms?

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,210 @@
+
+        // If this method has no uncovered branches, or it is the most-invoked method,
+        // use the default weight and skip this step.
+        if (covDet.uncoveredBranches != 0 && numSuccessfulInvocations != maxSuccessfulCalls) {
```
"maxSuccessfulCalls" and "numSuccessfulInvocations" are counting the same thing.  Please use consistent terminology (not "...Calls" vs "...Invocations").

Please make any other terminology consistent throughout the code, too.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,210 @@
+  /** How often to recompute branch coverage, in steps. */
+  private final int branchCoverageInterval = 100;
+
+  /** Maximum number of successful calls so far to any method under test. */
```
What is the definition of "successful calls" as opposed to "calls" or "selections"?

Is the word "successful" motivated by "succ(m)" in the GRT paper?  Please connect the dots for me.

huangwaylon replied:
Right, I originally used "successful" due to the GRT paper's definition of succ(m) as the number of "successful invokations" or a method. However, we concluded that the GRT paper didn't define how they defined "successful invokation", we decided to use the number of times a method is selected to construct a new and unique sequence for the definition of `succ(m)` and `maxSucc(M)`. I've rewritten usages of "successful calls/invokations" to "selections".

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,210 @@
+    stepNum += 1;
+    if (stepNum % branchCoverageInterval == 0) {
+      methodSelections.clear();
+      CoverageTracker.instance.collect();
```
What is the relationship between CoverageTracker.instance.collect() and updateWeightsForOperations() which is defined immediately below?  That was not clear to me.

mernst commented on src/main/java/randoop/main/GenInputsAbstract.java:
```diff
@@ -517,6 +517,14 @@ public Limits(int time_limit, int attempted_limit, int generated_limit, int outp
+   */
+  @Unpublicized
+  @Option("Enable bloodhound to prioritize methods with lower branch coverage.")
+  public static boolean enable_bloodhound = true;
```
This should be changed to `false` before merging -- unless we have completed experiments showing that Bloodhound is always superior, but I suspect we will merge before completing such experiments.

huangwaylon commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,199 @@
+      original.close();
+    } catch (IOException e) {
+      e.printStackTrace();
+      return null;
```
In regards to the instrumentation process, initially I was thinking that Randoop might want get as far as possible. The caller to this method `instrumentClass` is MemoryClassLoader in `loadClass` and it does check for a null result. Thus, if instrumentation fails, I wonder if it's a good idea to let the parent classloader attempt to load the class. Alternatively, should the class be loaded by the memory class loader, but left  un-instrumented if the instrumentation process fails? Or should we just terminate Randoop?

mernst replied:
What is the cause of the problem?  That is, when is the result null?
 * It it a user error or misconfiguration?  If so, it is something that the user can fix, and the user shoud do so.
 * Is it nondeterministic or for some reason you do not understand?

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,237 @@
+        numSuccessfulInvocations = 0;
+      }
+
+      // If this method has no uncovered branches, or it is the most-selected method,
```
Why?  The paper does not have a special case for the most-selected method, so this seems to differ from the paper's formula.
I suggest that you make this method as simple as possible, and make it obviously equivalent to the formula in the paper.

huangwaylon replied:
Hi Mike,
This extra clause is meant to catch the case where `uncovRatio(m) == 0` and `succ(m) / maxSucc(M) == 1` in which case `w(m, 0)` would evaluate to zero.
![image](https://user-images.githubusercontent.com/15660492/38847567-2d8067f8-41b7-11e8-8e95-f95bea29c4eb.png)
This is problematic because `Randomness.randomMemberWeighted` in Randoop does not allow an element's weight to be zero. I am thinking that it might make sense to treat this as an uncovered method and give it a default weight, or adjust random member selection to ignore elements with zero weight. Should I further document this choice or make the change?

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,237 @@
+    // Check that branch coverage details are available for this method.
+    if (covDet != null) {
+      // The number of successful invocations of this method. Corresponds to succ(m).
+      Integer numSuccessfulInvocations = methodInvocationCounts.get(operation);
```
Rename this variable to succM.  Your name is a better one, but it is more important to make this method correspond in an obvious way to the formula in the paper.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,237 @@
+        // Call ratio of this method. Corresponds to succ(m) / maxSucc(M) in the GRT paper.
+        double callRatio;
+
+        if (maxSuccessfulInvocationsOfAnyMethod == 0) {
```
I suggest initializing `maxSuccessfulInvocationsOfAnyMethod` to 1, in order to simplify this code.  (It won't need the check any more.)  You can document at its initialization that this is because it appears in the denominator of a division.

mernst commented on src/main/java/randoop/main/GenTests.java:
```diff
@@ -319,6 +318,11 @@ public boolean handle(String[] args) {
 
+    if (GenInputsAbstract.enable_bloodhound) {
+      // Copy the list of methods under test into Bloodhound's internal list.
+      Bloodhound.instance.initBloodhound(operations);
```
Why is Bloodhound an singleton that is created when Randoop is loaded, but then side-effected later?
This is a confusing design pattern.  I think it should be created with the set of operations, which should not be side-effected later.
Then the logic of `initBloodhound` can be moved into the Bloodhound constructor.
Even if you want to have an instance, the instance could be created here.

The same comment applies to `CoverageTracker`:  put the logic of `setClassesUnderTest` in the constructor.
Can CoverageTracker be a field of Bloodhound?  Encapsulating the logic within Bloodhound would keep Randoop simpler and more encapsulated, if it is possible.

Also, can you combine the two occurrences of 
```
    if (GenInputsAbstract.enable_bloodhound) {
```
?  Putting the logic together will be cleaner, I think.
Or is that impossible because `createModel` loads the classes?  In that case, document it.

mernst commented on src/main/java/randoop/generation/ForwardGenerator.java:
```diff
@@ -93,6 +96,13 @@ public ForwardGenerator(
+
+    if (GenInputsAbstract.enable_bloodhound) {
+      // If Bloodhound is enabled, select the next operation while considering the methods' weights.
+      this.operationSelector = Bloodhound.instance;
```
Shouldn't this create the Bloodhound instance, just as it creates the UniformRandomMethodSelection instance two lines later?

mernst commented on src/main/java/randoop/generation/InstrumentingClassLoader.java:
```diff
@@ -0,0 +1,51 @@
+
+    // Attempt to instrument the class identified by the class name.
+    final byte[] bytes = coverageTracker.instrumentClass(name);
+    if (bytes != null) {
```
Under what circumstances is `bytes` null?  Does this indicate some problem with the coverage tracker?  Or is it a normal occurrence?  Please add a brief comment.

huangwaylon replied:
I've documented this as: The byte array will be null for a class that is not either explicitly under test or is not a nested class of a class that is explicitly under test.

mernst commented on src/main/java/randoop/generation/TypedOperationSelector.java:
```diff
@@ -0,0 +1,11 @@
+ * new test sequence.
+ */
+public interface TypedOperationSelector {
+  public abstract TypedOperation selectOperation();
```
Please add Javadoc for this method.  (It can be short, but it should exist.)

mernst commented on src/main/java/randoop/generation/UniformRandomMethodSelection.java:
```diff
@@ -0,0 +1,29 @@
+  private final List<TypedOperation> operations;
+
+  /**
+   * Saves a reference to the list of methods under test.
```
The main thing this does is to create a new object.  Please reflect that in the documentation.

mernst commented on src/main/java/randoop/generation/UniformRandomMethodSelection.java:
```diff
@@ -0,0 +1,29 @@
+   * @param operations methods under test
+   */
+  public UniformRandomMethodSelection(List<TypedOperation> operations) {
+    this.operations = operations;
```
Should this make a defensive copy?  (I don't know the answer -- I'm asking.)

huangwaylon replied:
For some reason, making a copy of the list causes a few system tests to fail - specifically the ones that check coverage goals.  I noted in the documentation the choice to not make a copy.

mernst commented:
I made some changes to master that cause trivially-resolvable conflicts.  Please make all changes based on code review before pulling master into this branch.  Thanks!

huangwaylon commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,237 @@
+    // The number of methods under test, corresponds to |M| in the GRT paper.
+    int numOperations = this.operationSimpleList.size();
+
+    // Default weight is uniform probability.  It is used if no coverage details are available.
```
Hi Mike,
The paper does not seem to indicate what value should be used for methods with no coverage information. It appears that 1/numOperations will become relatively large as `k` increases. However, there is a lower bound on the multiplier to w(m, 0), namely the `1/ln(|M| + 3)`. I wonder if it's a better choice to set `w(m, k) = 1/ln(|M| + 3)` for all `m` that is uncovered?
![image](https://user-images.githubusercontent.com/15660492/38847324-e6067d0a-41b5-11e8-8229-7d9dc5213fc4.png)

mernst commented on src/main/java/randoop/util/Randomness.java:
```diff
@@ -171,18 +172,22 @@ public static int nextRandomInt(int i) {
 
+    // Subset of elements from the input list whose weights are nonzero. The selected element
+    // will be chosen from this list.
+    List<T> listOfElementsWithNonZeroWeights = new ArrayList<>();
```
This code was already very inefficient.  It's crazy to perform an O(n) traversal, creating a whole new array just for the purpose of running an O(log n) binary search on it.

Suppose there are 10,000 methods under test.  Then the old code created a 10,000-element array every time it wanted to select a method.  The new code is slightly worse, because an ArrayList is more expensive than an array.

Neither an array nor an ArrayList is necessary.
It's enough to
 * determine the total weight -- this can be maintained by the client and passed in.
 * choose a value that is no more than that, and
 * walk through the elements one-by-one, maintaining a running weight and return the element that makes the total weight greater than the chosen value.

The big benefit is that there is no need to allocate an array or ArrayList.  A smaller benefit is that on average, only half of the elements need to be touched before one can be returned.

I will create another pull request to change Randomness.
(These changes to Randomness need to be undone, in order to avoid conflicts.)

The second part is to change Bloodhound to cache and update the total weight, which will improve efficiency.

huangwaylon commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,235 @@
+    this.coverageTracker = coverageTracker;
+    // Compute an initial weight for all methods under test. We also initialize the uncovered ratio
+    // value of all methods under test by updating branch coverage information. The weights for all methods may not be uniform
+    // in cases where we have methods with "zero" branches and methods with non-"zero" branches.
```
A method will have zero branches, according to Jacoco, if it is only composed of straight line code that will entirely executed as a sequence. 
In testing Bloodhound, I found that getMyInt and the MyClass constructor from https://github.com/huangwaylon/grt_tests_for_randoop/blob/master/MyClass.java both had "zero" branches.

mernst replied:
(My guess was that a straightline program would have 1 branch.)
So, there are no methods with 1 branch?  A method with a single if statement would have 2 branches?
If you could document this in the code, that would be helpful to future maintainers.
Thanks!

huangwaylon commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,217 @@
+   * Attempts to instrument a class and return the byte array representation.
+   *
+   * @param className name of the class to instrument
+   * @return the instrumented bytes of the class, or null if the class does not need to be
```
I agree with this change in that an exception will always be thrown if instrumentation fails!

huangwaylon commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,235 @@
+   */
+  private void updateBranchCoverageMaybe() {
+    if (totalSuccessfulInvocations % branchCoverageInterval == 0) {
+      methodSelectionCounts.clear();
```
I think that it could be, however `methodSelectionCounts` belongs to Bloodhound and `summarizeCoverageInformation` belongs to CoverageTracker. To clear the map, would it be a good idea to pass the methodSelectionCounts map into summarizeCoverageInformation?

huangwaylon replied:
Alternatively, did you mean to abstract out the two lines into a separate method that would both clear the methodselectionscount map and call summarizeCoverageInformation on the coverageTracker instance?

huangwaylon commented on src/main/java/randoop/generation/InstrumentingClassLoader.java:
```diff
@@ -0,0 +1,54 @@
+    // class that is under test.
+    if (bytes != null) {
+      // Use the instrumented bytes to define the class.
+      loadedClass = defineClass(name, bytes, 0, bytes.length);
```
Here InstrumentingClassLoader makes a call to defineClass instead of loadClass. I didn't see a definition of defineClass that makes use of the resolve flag which is why I left it off. Additionally, the example from which this class was originally derived from did not make use of the loadClass flag: http://www.jacoco.org/jacoco/trunk/doc/examples/java/CoreTutorial.java

mernst replied:
Here is the documentation for `loadClass`: https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html#loadClass(java.lang.String,%20boolean)
It says to call `resolveClass` if the boolean is true, and it also says that calling `findClass` is better.
At the least, if the code doesn't use the `resolve` variable, I think this should be documented.  But, even better would be to try to make the suggested call and see if anything breaks.
Or is that undesirable because it would prevent further instrumentation?

huangwaylon replied:
For the if case, define class is needed because the class is "defined" from an array of bytes. (These bytes are created from instrumenting a class under test".

The ClassLoader documentation notes that:
"However, some classes may not originate from a file; they may originate from other sources, such as the network, or they could be constructed by an application. The method defineClass converts an array of bytes into an instance of class Class. Instances of this newly defined class can be created using Class.newInstance."

Alternatively, did you want InstrumentingClassLoader to override `findClass` instead of `loadClass`?

huangwaylon replied:
I think your comment on "preventing further instrumentation" is right if we override `findClass` instead of `loadClass`. 
The documentation of ClassLoader notes that:
"Finds the class with the specified binary name. This method should be overridden by class loader implementations that follow the delegation model for loading classes, and will be invoked by the loadClass method after checking the parent class loader for the requested class. The default implementation throws a ClassNotFoundException."
Thus, the ClassLoader would find an uninstrumented version of the class in the parent classloader and return a `Class<?>` that is not instrumented.

huangwaylon commented on src/main/java/randoop/reflection/OperationModel.java:
```diff
@@ -90,6 +91,9 @@
   private List<Pattern> omitMethods;
 
+  /** Coverage tracker used to instrument and load classes under test. */
+  private static CoverageTracker coverageTracker;
```
The coverageTracker for both classes is actually a reference to an instance of CoverageTracker that is created in GenTests. OperationModel needs the coverage tracker to instrument classes that are loaded in. Bloodhound is initialized only after OperationModel finishes and uses the coverage tracker to collect branch coverage information.

huangwaylon commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,217 @@
+
+  /**
+   * Branch coverage for a single method under test. Records "uncovRatio" which is the ratio of
+   * uncovered branches to total branches. In cases where total branches is zero, the uncovered
```
I've added this line to explain:
A method with only straight-line code that will always be executed entirely as a sequence, with no possibility of an exception or early return, will have zero "total branches".

huangwaylon commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  /**
+   * Summarizes coverage information for all methods under test. At this point, Jacoco has already
```
Jacoco inserts "probes" into the instrumented classes to gather coverage information. This coverage information is updated immediately as the methods in these classes are executed. However, the CoverageTracker's view of the coverage information isn't updated unless we explicitly ask Jacoco for the information. Thus, I named it "summarize" because new information isn't being computed/created, rather we're summarizing the information given to us by Jacoco.  I do agree that updateBranchCoverageMap is more clear though.

huangwaylon commented:
Hi Mike,

I've determined that `resolve` is always false. Therefore, adding if(resolve){resolveClass(class)}; has no effect.
This is actually because CoverageTracker calls, on line 138:
`return instrumentingClassLoader.loadClass(className);`
And `loadClass` which takes in only a `String` is defined as:

> public Class<?> loadClass(String name) throws ClassNotFoundException {
>         return loadClass(name, false);
> }
Thus, this would imply that your second hypothesis is likely correct in that resolution automatically occurs at some later step without our need to explicitly do so.

huangwaylon commented:
Randoop with Bloodhound now uses Jacoco as a Java agent to instrument classes. It then makes calls to Jacoco's library that allows us to retrieve the coverage information of classes that are under test.

When Bloodhound is enabled, we also need to include Jacoco on the class path and also specify it on the Xbootclasspath and javaagent arguments.

For instance, the command for the example suite that I've been working with (https://github.com/huangwaylon/grt_tests_for_randoop) is:

`java -ea -cp .:randoop.jar:replacecall.jar:jacocoagent.jar -Xbootclasspath/a:/home/waylonh/test_digdog/replacecall.jar -Xbootclasspath/a:/home/waylonh/test_digdog/jacocoagent.jar -javaagent:replacecall.jar=--replacement-file=/home/waylonh/test_digdog/replacement_file.txt -javaagent:jacocoagent.jar randoop.main.Main gentests --classlist=classlist.txt --output-limit=1000 --enable-bloodhound=true`

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,302 @@
+      // - Object.<init> and Object.getClass which Randoop always includes as methods under test.
+      // - Classes that are from the JDK (eg, java.lang) or from external jars.
+      // - Getters and setters operations for public member variables that Randoop synthesizes.
+      // - Abstract method declarations.
```
Also interface methods?

mernst commented on src/main/java/randoop/generation/AbstractGenerator.java:
```diff
@@ -331,6 +331,13 @@ public void createAndClassifySequences() {
+            // Increment the number of times this last operation was successfully invoked
+            // as the final method of a new sequence.
+            TypedOperation lastOperation = eSeq.sequence.getOperation();
+            newRegressionTestHook(lastOperation);
```
Both the invocation and the definition of this method are guarded by `enable_bloodhound`.  That doesn't achieve abstraction.  At the least, the call here should be unconditional, and the hook method should take the sequence rather than the operation.

Ideally, when enable_bloodhound is true, Randoop could set a field to an object, like TypedOperationSelector, that does the work, and enable_bloodhound wouldn't be used except at Randoop startup.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,302 @@
+ *
+ * <p>Bloodhound computes a weight for each method under test by taking a weighted combination of
+ * the uncovered branch ratio and the ratio between the number of times the method has been
+ * successfully invoked (to be the last statement of a new regression test) and the maximum number
```
Our definition is about the number of times the method was selected for a test, and the test was a regression test.
Another possible definition would be the total number of times the method appears in any regression test.
Both are consistent with the paper.  I think what we have is more likely to be what the authors intended, but it's impossible to know for sure.

mernst replied:
My above comment is probably worth adding to the documentation.

huangwaylon replied:
Documentation now includes comment on the two definitions of "successful invocations".

huangwaylon commented:
Fixed error where nested classes had no coverage information. After making the switch to using Jacoco as a Java agent, Bloodhound didn't have a complete list of all the classes under test, just the ones explicitly listed by the user. Now passing in a complete list of all the classes under test so that Bloodhound can analyze the coverage information of nested classes as well.

mernst commented on src/testinput/java/java7/util7/Collections.java:
```diff
@@ -590,7 +591,7 @@ private static void swap(Object[] arr, int i, int j) {
             throw new IndexOutOfBoundsException("Source does not fit in dest");
 
         if (srcSize < COPY_THRESHOLD ||
-            (src instanceof RandomAccess && dest instanceof RandomAccess)) {
```
What is the reason for the many whitespace changes in this file?
We should keep this file in sync with the actual JDK, to make comparisons easier in the future.
Does the new or the old indentation correspond to the JDK?

mernst commented on src/testinput/java/java7/util7/Collections.java:
```diff
@@ -29,6 +29,7 @@
 import java.io.IOException;
 import java.lang.reflect.Array;
 import java.util.Iterator;
+import java.util.Random;
```
Why is this added?

mernst commented on src/systemtest/java/randoop/main/RandoopSystemTest.java:
```diff
@@ -192,16 +192,17 @@ public void runCollectionsTest() {
         new CoverageChecker(
             options,
             "java7.util7.Collections.asLifoQueue(java7.util7.Deque) exclude",
-            "java7.util7.Collections.binarySearch(java7.util7.List, java.lang.Object) exclude",
```
There seem to be a lot of changes to coverage goals to accommodate bloodhound.  But bloodhound is not enabled by default.  So why are these needed?  Are they only needed when bloodhound is enabled?  Should they be in the pull request, or delayed until a future pull request makes bloodhound the default (if that ever happens)?

huangwaylon replied:
A subset of these changes to coverage goals were due to Bloodhound. I reverted the RandoopSystemTest class to be the same as Master and observed that a few system tests were failing even with Bloodhound disabled. I believe I have pinpointed the problem to UniformRandomSelector which makes a copy of the list of methods under test. Having a direct reference allows these tests to pass locally. I have pushed the changes and waiting to confirm that it passes on Travis as well.

mernst commented on src/main/java/randoop/generation/AbstractGenerator.java:
```diff
@@ -331,6 +331,10 @@ public void createAndClassifySequences() {
         } else {
           outRegressionSeqs.add(eSeq);
+
+          // Bloodhound uses this hook to increment the number of times the last operation
```
I don't think this comment belongs in AbstractGenerator.  Maybe in Bloodhound's implementation istead?

mernst commented on src/main/java/randoop/generation/CoverageTracker.java:
```diff
@@ -0,0 +1,172 @@
+
+      // The reader requires a session info visitor, however we don't need any information from it.
+      reader.setSessionInfoVisitor(
+          new ISessionInfoVisitor() {
```
Minor:  I would find it clearer to create a private singleton class `DummySessionInfoVisitor` and use it here.

mernst commented on src/main/java/randoop/generation/Bloodhound.java:
```diff
@@ -0,0 +1,384 @@
+    Double uncovRatio = coverageTracker.getBranchCoverageForMethod(methodName);
+
+    if (uncovRatio == null) {
+      // Default to zero for methods with no coverage information.
```
"Default to zero" is inconsistent with `uncovRatio = 0.5;` on line 303.  Which one is desired?
